### ISSUE TITLE
Memory leaks when using tf.keras.metrics update_states in multi-threads.

### ISSUE THREAD

**Issue by workingloong** [tensorflow/tensorflow#35044]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.0.0
- Python version: 3.6.8
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**
The memory increases and does not release when I use ` tf.keras.metrics.update_states` in multi-threads on Linux. I view the used memory by "memory_profiler" and find that the memory increase after each thread starts.

![image](https://user-images.githubusercontent.com/18071380/70680739-150c6600-1cd4-11ea-9f04-6a1724b72d23.png)

**Describe the expected behavior**
The memory increases greatly after only the first thread starts.

**Code to reproduce the issue**
```python
import threading
import time
import tensorflow as tf
import numpy as np
from memory_profiler import profile

auc = tf.keras.metrics.AUC()
lock = threading.Lock()


def update_metric(metric):
    labels = np.random.randint(0, 2, 20000)
    preds = np.random.random(20000)
    metric.update_state(labels, preds)


def update_metric_mini(metric):
    for i in range(10):
        labels = np.random.randint(0, 2, 2000)
        preds = np.random.random(2000)
        metric.update_state(labels, preds)


def test_mem(metric):
    update_metric(metric)
    # update_metric_mini(metric)


def test_tf_metric():
    with lock:
        test_mem(auc)


@profile
def test_threads():
    t1 = threading.Thread(target=test_tf_metric)
    t2 = threading.Thread(target=test_tf_metric)
    t3 = threading.Thread(target=test_tf_metric)
    t1.start()
    time.sleep(0.1)
    t2.start()
    time.sleep(0.1)
    t3.start()
    time.sleep(0.1)
    t1.join()
    t2.join()
    t3.join()
    time.sleep(1)
    print("end ")


test_threads()
```
In this test code snippet, if we use the `update_metric_mini` which split the updated outputs and labels to small chunks and update metric with those chunks, the memory increases small too. I wonder whether the `update_states` in each thread creates global variables (e.g. placeholder) for input arguments `outputs` and `labels` and the variable size is proportional to the size of input arguments.

***other info*
We encounter this issue when developing evaluation in [ElasticDL](https://github.com/sql-machine-learning/elasticdl) and resolving the ElasticDL [issue 1568](https://github.com/sql-machine-learning/elasticdl/issues/1568)

**Comment by amahendrakar** [comment]
> The memory increases greatly after only the first thread starts.

@workingloong,
Is this still an issue?

On running the code with TF v2.3, I did not observe huge memory growth after the start of the first thread. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/da01206f8991ef773c161f7cbe437748/35044.ipynb). Thanks!

**Comment by workingloong** [comment]
> > The memory increases greatly after only the first thread starts.
> 
> @workingloong,
> Is this still an issue?
> 
> On running the code with TF v2.3, I did not observe huge memory growth after the start of the first thread. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/da01206f8991ef773c161f7cbe437748/35044.ipynb). Thanks!

I don't find `test.py` to execute `!python test.py` in this Ipython notebook. I have tested the code in Linux with TF v2.3 and it is still an issue.

![image](https://user-images.githubusercontent.com/18071380/91514726-09240b00-e91a-11ea-99c7-f3e3cd12e9ab.png)

**Comment by jvishnuvardhan** [comment]
@workingloong Is this still an issue for you? Can you please test with recent `TF2.7` and `tf-nightly` and let us know whether the issue persists with recent TF version. Also, share a complete standalone code to reproduce the issue. Thanks!
